[18기 남궁선아] 모달창 로그인/회원가입 자체 로그인 구현완료, 카카오 소셜로그인 구현완료, nav바 연결 완료

sunaaank opened the pull request:
## 수정 사항 간략한 한줄 요약
- `로그인/회원가입 모달` 레이아웃 작업완료
- `로그인/회원가입 모달` 함수형 컴포넌트로 기능 구현 완료
- `nav바`와 `모달창` 연결 및 `nav바` 스크롤 이벤트 및 기능 구현 완료
		 
## 수정 사항들 자세한 내용
21.03.31
- `로그인 모달` 레이아웃 작업중
- `로그인 모달` 하위 컴포넌트화 제작 완료, props 작업중

21.04.01
- `로그인/회원가입 모달` 레이아웃 작업 완료
- `로그인/회원가입 모달` 로그인 상태에 따라 모달창 조건부렌더링 구현 완료
- `로그인/회원가입 모달` 기본 기능 구현 완료(input 상태 관리 구현 완료, 리셋 구현)

21.04.02
- `카카오 소셜로그인` API 통신 완료
- `Git rebase` 이슈 해결 완료
- 코드리뷰 반영 `리팩토링` 진행(함수 재사용, 불필요한 코드 삭제)

21.04.03.
- 코드리뷰 반영 `리팩토링` 진행(조건부렌더링)
- `모달창` z-index 최상위를 위한 글로벌 스타일 relative 속성 적용
- `Footer` 공통 컴포넌트 `fixed` -> `absolute`로 변경, `height`값 적용
- `자체 로그인/회원가입` 데이터 통신 준비완료
- `카카오 소셜로그인` 컴포넌트 재사용, 기능 적용 완료

21.04.04.
- `input` 값 조건에 따른 오류메세지 레이아웃 조건부렌더링 및 기능 구현(id, pw 정규식 사용)
- `자체 로그인/회원가입` 버튼 클릭 시 전송 후 이전 페이지로 이동 기능 구현 (`history.goBack()`) 

21.04.05.
- `nav바` 로그인 여부에 따라 버튼 상태 변경 기능
- `nav바` 로그인/회원가입 버튼에 따라 다른 모달창 on 기능 
- `모달창` 바깥(dimmed 영역) 클릭 시 모달 off 기능
- `모달창` dimmed 처리 후 `Form`, `FormContainer` 영역 이벤트버블링 방지

21.04.06.
- `nav바` 스타일 변경 스크롤 이벤트 구현완료
- `nav바` 관련 state 조작 로직 최종 구현 및 테스트 완료 
- 백엔드와 `자체 로그인/회원가입`, `카카오 소셜로그인` API 최종 테스트 완료

## 기타 질문 및 특이 사항
- 추가구현예정1 : 로컬스토리지 토큰 저장 후 토큰 값 여부에 따라 작동하도록 로직 변경 
- 추가구현예정2 : `모달창` 바깥(dimmed 영역) 스크롤 막는 기능

💌 질문 (해결완료) 💌
**: 회원가입/로그인 `input`창의 유효성을 조건에 따라 검사하는 함수인 `isInputValid`의 작명에 대해 질문드립니다.** 
✅ 처음에는 조건에 해당되면 `true`를 반환하라는 아래와 같은 함수를 썼습니다. (이전에 쓴 코드라 현재 코드에 작성된 조건보다는 간단한 상태입니다) 이에 따라 `input`창에 경고메세지가 뜨게 하는 코드에서도 `IsInputValid`라고 조건을 걸어 작성했습니다.
```jsx
const IsInputValid = item => {
    if (item === 'email') {
      return inputValue.email && !isIdValid.test(inputValue.email);
    } else if (item === 'password') {
      return inputValue.password && inputValue.password.length < 10;
    }
  };

...
{input.type === 'email' && IsInputValid('email') && (
  <FormValidChk>정확하지 않은 이메일입니다.</FormValidChk>
 )}
```
✅ 그런데 이렇게 함수를 작성할 경우, 유효한 값인데 `false`를 반환하는 로직이라 다른 사람이 코드를 읽을 때 직관적으로 이해가 어렵다는 걸 알게 되어 지금의 코드로 유효한 값이면 `true`를 반환하도록 리팩토링하고, 추후 함수를 사용할 때도 `!IsInputValid`라고 작성하였습니다.
✅ 가장 간단한 리팩토링 방법은 변수명을 바꿔주는 걸텐데 ㅎㅎㅎ돌아보니 돌아돌아 왔네요......😂💦 마땅한 이름이 생각이 안납니다...........추천받습니다.......



## 체크 리스트 (아래 사항들이 전부 체크되어야만 merge가 됩니다!)
- [x] 필요한 test들을 완료하였고 기능이 제대로 실행되는지 확인 하였습니다.
- [x] Wecode의 코드 스타일 가이드에 맞추어 코드를 작성 하였습니다.
- [x] 제가 의도한 파일들과 수정 사항들만 커밋이 된 것을 확인 하였습니다.
- [x] 본 수정 사항들을 팀원들과 사전에 상의하였고 팀원들 모두 해당 PR에 대하여 알고 있습니다.
- [x] Git rebase와 squash를 했고 커밋 수가 하나 인것을 확인 했습니다.

JHyeon0 commented on src/Components/Form/FormLayout.js:
```diff
@@ -0,0 +1,42 @@
+import React from 'react';
+import styled from 'styled-components';
+
+export default function FormLayout({ children }) {
+  return (
+    <FormContainer>
+      <FormHeader>
+        <FormLogo>
+          <Logo />
+        </FormLogo>
+      </FormHeader>
+      {children}
+    </FormContainer>
+  );
+}
```
세션 내용 바로 적용하셨네요! 좋습니다!
👍 👍 👍 👍

JHyeon0 commented on src/Components/Form/FormLayout.js:
```diff
@@ -0,0 +1,42 @@
+  margin: 0 auto;
+`;
+
+const Logo = styled.img.attrs({
+  alt: '왓더피디아 로고',
+  src: '/images/logo.png',
+})`
```
alt와 src 속성을 스타일드 컴포넌트에 넣은 이유가 있을까요?

sunaaank replied:
attrs를 써보는 연습해보기 + jsx를 깔끔하게 쓰려고 분리해서 작성해보았습니다. 
스타일드 컴포넌트에서 attrs속성은 props로 변경되는 데이터값을 줄 때만 사용하기를 권장되는지, 아니면 고정값에 사용하여도 무방한 지 궁금합니다!

아래 페이지를 참고했습니다.
https://stackoverflow.com/questions/56378786/what-is-the-use-case-of-styled-components-attrs-function

JHyeon0 commented on src/Components/Form/Form.js:
```diff
@@ -0,0 +1,112 @@
+        ))}
+      </div>
+      <Button value={format} color={format} onReset={onReset} />
+      {format.type === 'signUp' && (
+        <FormP>
+          <FormSpan>
+            이미 가입하셨나요?
+            <FormSpanClick onClick={handleLogin}>로그인</FormSpanClick>
+          </FormSpan>
+        </FormP>
+      )}
+
+      {format.type === 'login' && (
+        <FormP>
+          <FormSpanClick>비밀번호를 잊어버리셨나요?</FormSpanClick>
+          <FormSpan>
+            계정이 없으신가요?
+            <FormSpanClick onClick={handleLogin}>회원가입</FormSpanClick>
+          </FormSpan>
+        </FormP>
+      )}
```
회원가입 로그인 두 부분에서
통째로 조건부 렌더링말고 필요한 부분만 조건부렌더링 하면 더 좋을 것 같네요 :)

sunaaank replied:
네, 부분별 조건부렌더링으로 변경하였습니다!✍

JHyeon0 commented on src/Components/Modal/Modal.js:
```diff
@@ -0,0 +1,113 @@
+    password: '',
+  });
+
+  const handleLogin = () => {
+    setIsLogin(!isLogin);
+    isLogin
+      ? setLoginInput({ ...loginInput, email: '', password: '' })
+      : setSignInput({ ...signInput, name: '', email: '', password: '' });
+  };
+
+  const handleChange = e => {
+    const { name, value } = e.target;
+    isLogin
+      ? setLoginInput({ ...loginInput, [name]: value })
+      : setSignInput({ ...signInput, [name]: value });
+  };
+
+  const onReset = () => {
+    setLoginInput({ ...loginInput, email: '', password: '' });
+    setSignInput({ ...signInput, name: '', email: '', password: '' });
+  };
```
`handleLogin`과 `onReset`에서 초기화 시키는 부분의 코드가 중복되고 있네요!
또한 불필요한 스프레드가 사용되고 있는 것 같습니다.
```jsx
setLoginInput({ ...loginInput, email: '', password: '' }); // 불필요 spread
setLoginInput({ email: '', password: '' }); // 이렇게!

setSignInput({ ...signInput, name: '', email: '', password: '' }); // 불필요 spread
setSignInput({ name: '', email: '', password: '' }); // 이렇게!
```
 
```jsx
const resetInput = (type) => {
  // 로그인인지, 회원가입인지에 따라서 초기화
}
```
`resetInput` 함수를 여러 곳에서 활용할 수 있겠네요 :)

sunaaank replied:
불필요한 코드는 삭제하고, 함수재사용하는 방식으로 변경하였습니다. 감사합니다🌳

sunaaank replied:
불필요한 코드는 삭제하고, 함수재사용하는 방식으로 변경하였습니다. 감사합니다🌳

JHyeon0 reviewed (commented):
선아님 고생하셨습니다!

컴포넌트 재사용, 스타일드 컴포넌트 등등 너무 잘 하고 계시네요 :)
화이팅 화이팅!!

sunaaank committed:
로그인 레이아웃 완료, 소셜로그인 기능,  기본로그인 구현 완료

Fix : refactor 코드리뷰 반영

Fix : Footer 레이아웃 수정

Fix : Modal 최상위 z-index를 위한 글로벌스타일 position:relative 적용

Add : 자체 로그인/회원가입 데이터 통신 준비완료, 카카오소셜로그인 연결 완료

Fix : refactor 조건부렌더링 코드리뷰 반영

Fix : refactor 모달창 Form 조건부렌더링

Add : 로그인/회원가입 조건 필터링 기능구현

Add : 모달창 제출 시 뒤로가기 기능

Fix : state 위치 변경

Fix : 검색바 아이콘 레이아웃 수정

Add : 비밀번호 정규식 조건 추가

Fix : Refactor : 코드리뷰 반영, 정규식체크 오타 수정

Add : 모달창 dimmed영역 제어 이벤트 구현완료, 네비게이션바 모달창 연결

Fix : 모달창 off 이벤트버블링 방지

Add : Nav 스크롤 이벤트 적용

Fix : 모달창 위치 nav바 위로 조정

Fix : console 삭제

Fix : 카카오 로그인 API 키값 재조정중

JHyeon0 commented:
선아님! 코드 리뷰는 아니지만, 오늘 오프라인에서 주신 질문이 PR의 질문에 이미 잘 되어 있어서 바로 답변 드리겠습니다 :)
질문 너무 정리 잘 되어 있어서 답변하기가 너무 편하네요 👍

1. 만약 함수명만 바꾸실 거라면 `isInvalidInput` 으로 바꾸면 될 것 같네요.
2. 두 번째 방법은 아래 코드로 설명드리겠습니다.
먼저 현재 선아님이 올리신 PR의 함수입니다.
```js
const isValidInput = item => {
    if (item === 'email') {
      if (inputValue.email && !isIdValid.test(inputValue.email)) {
        return false;
      } else {
        return true;
      }
    } else if (item === 'password') {
      if (inputValue.password && !isPwValid.test(inputValue.password)) {
        return false;
      } else {
        return true;
      }
    } else if (item === 'name') {
      if (inputValue.name && inputValue.name.length < 2) {
        return false;
      } else {
        return true;
      }
    }
  };
```

아래는 수정한 함수입니다. 객체의 각 프라퍼티가 true, false 값입니다!

```js
const isValidInput = item => {
  const INPUT_VALIDATION_MAP = {
    email: inputValue.email && isIdValid.test(inputValue.email),
    password: inputValue.password && isPwValid.test(inputValue.password),
    name: inputValue.name && inputValue.name.length>1
  };

  return INPUT_VALIDATION_MAP[item];

};
```
